src/host/alloc_cache.jl

```diff
@@ -0,0 +1,157 @@
+using ..GPUArrays
+
+@static if VERSION < v"1.11"
+    using ScopedValues
+else
+    using Base.ScopedValues
+end
+
+mutable struct AllocCache{T <: AbstractGPUArray}
+    lock::ReentrantLock
+    busy::Dict{UInt64, Vector{T}} # hash(key) => GPUArray[]
+    free::Dict{UInt64, Vector{T}}
```

Comment:
Why lock down the type here? `T=CuArray` is going to result in insufficiently typed values anyway. Wouldn't it be better to make it a `Dict{UInt64, Any}` so that the user doesn't have to specify the array type?

Reply:
Good point. We are specifying the array type in CUDA/AMDGPU/etc ctor anyway